src/support/allocators/node_allocator.h

```diff
@@ -0,0 +1,406 @@
+ * @brief Efficient allocator for node-based containers.
+ *
+ * The combination of Allocator and MemoryResource can be used as an optimization for node-based
+ * containers that experience heavy load.
+ *
+ * ## Behavior
+ *
+ * MemoryResource mallocs pools of memory and uses these to carve out memory for the nodes. Nodes
+ * that are destroyed by the Allocator are actually put back into a free list for further use. This
+ * behavior has two main advantages:
+ *
+ * - Memory: no malloc control structure is required for each node memory; the free list is stored
+ * in-place. This typically saves about 16 bytes per node.
+ *
+ * - Performance: much fewer calls to malloc/free. Accessing / putting back entries are O(1) with
+ * low constant overhead.
+ *
+ * There's no free lunch, so there are also disadvantages:
+ *
+ * - It is necessary to know the exact size of the container internally used nodes beforehand, but
+ * there is no standard way to get this.
+ *
+ * - Memory that's been used for nodes is always put back into a free list and never given back to
+ * the system. Memory is only freed when the MemoryResource is destructed.
+ *
+ * - The free list is a simple last-in-first-out linked list, it doesn't reorder elements based on
+ * proximity. So freeing and malloc'ing again can become a random access pattern which can lead to
+ * more cache misses.
+ *
+ * ## Design & Implementation
+ *
+ * Allocator is a cheaply copyable, `std::allocator`-compatible type used for the containers.
+ * Similar to `std::pmr::polymorphic_allocator`, it holds a pointer to a memory resource.
+ *
+ * MemoryResource is an immobile object that actually allocates, holds and manages memory. Currently
+ * it is only able to provide optimized alloc/free for a single fixed allocation size which is given
+ * in the constructor. Only allocations that match this size will be provided from the preallocated
+ * pools of memory; all other requests simply use ::operator new(). To get the correct node size in
+ * all cases it is unfortunately necessary to have a look at the various implementations of
+ * std::unordered_map. There is currently no standard way of getting the node size programmatically.
+ *
+ * Node size is determined by memusage::NodeSize and verified to work in various alignment scenarios
+ * in `node_allocator_tests/test_allocations_are_used`.
+ *
+ * ## Further Links
+ *
+ * @see CppCon 2017: Bob Steagall “How to Write a Custom Allocator”
+ *   https://www.youtube.com/watch?v=kSWfushlvB8
+ * @see C++Now 2018: Arthur O'Dwyer “An Allocator is a Handle to a Heap”
+ *   https://www.youtube.com/watch?v=0MdSJsCTRkY
+ * @see AllocatorAwareContainer: Introduction and pitfalls of propagate_on_container_XXX defaults
+ *   https://www.foonathan.net/2015/10/allocatorawarecontainer-propagation-pitfalls/
+ */
+namespace node_allocator {
+
+namespace detail {
+
+/**
+ * In-place linked list of the allocations, used for the free list.
+ */
+struct FreeList {
+    void* next;
```

Comment:
Reading through the code again I now think that I actually have undefined behavior: I'm basically `reinterpret_cast`ing `T*` to `FreeList*`, and both types are completely unrelated, so according to the type aliasing rules accessing this is undefined behavior:

> Performing a class member access that designates a non-static data member or a non-static member function on a glvalue that does not actually designate an object of the appropriate type - such as one obtained through a reinterpret_cast - results in undefined behavior

See https://en.cppreference.com/w/cpp/language/reinterpret_cast

This can be fixed by using `memcpy`, and then I can also relax the alignment.

Reply:
I guess I didn't have enough coffee yet. I can't use reinterpret_cast, but since the `T` was already destroyed, I can use placement new in that memory to create the `FreeList`. No reinterpret_cast needed at all.

Reply:
`MemoryResource::Deallocate` now uses pacement new in e2a5547f12e5da2718ba17ebac5d5aeb5a2e24d5, and uses the returned pointer.